io.streams: multiple related changes

stepancheg opened the pull request:
- write syscall may return before all data written, handle that
- add explicit InputStream, OutputStream protocols
- move byte vector as output stream into submodule
- InputStream and OutputStream to read and write to memory
- RawFilePointer is shared pointer to RawFile. Necessary when multiple objects need to safely reference single RawFile
- reimplement io.files.easy.File with Buffed*Stream and RawFilePointer

stepancheg committed:
io.files.raw: move raw file implementations into platform submodule

stepancheg committed:
io.files.raw: RawFilePointer

Shared pointer to RawFile that mimics RawFile.

stepancheg committed:
io.streams.buffered: BuffededInputStream, BufferedOutputStream

BufferedOutputStream has flushOnNewlines? flag. When flag is on,
BufferedOutputStream scans input buffer for presence of '\n' character
and flushes content if newline is present.

Note, BufferedOutputStream does not flush buffer on destroy. See
explanation in comments.

stepancheg committed:
io.streams: write syscall can return before all data written

`write` syscall can return before all data written, for instance, when
write is interrupted or when descriptor is non-blocking.

This patch adds `writeBytes` function. `writeBytes` returns number of
bytes written, stream implementations must overload `writeBytes` now.

`write` functions return no values. `write(stream, buf, b)` now calls
`writeBytes` in loop until all data written.

Also add signature to `flush` function.

stepancheg committed:
io.files.easy: reimplement File using io.streams.buffered

stepancheg committed:
io.streams.memory: streams to read and write to memory

stepancheg committed:
io.streams: InputStream, OutputStream protocols

InputStream is readable stream of bytes. OutputStream is writable stream
of bytes. RawFile conforms to both these protocols.

stepancheg committed:
io.streams: move byte vector stream into its own module

and add some tests

jckarter commented:
On Unix at least, I believe `write` will only return a partial write if the fd is a nonblocking socket. Could byte streams also support the iterator and push protocols so they're interchangeable with byte sequences?

stepancheg committed:
io.streams: implement collection protocols

* InputStream is Iterator[Byte]
* OutputStream has push(stream, byte) and push(stream, string) functions

stepancheg committed:
io.streams: change readChar and readByte functions to return Maybe

stepancheg committed:
io.streams: writeByte shortcut

stepancheg committed:
io.streams: use ByteLike? and Integer? in `read` and `write` operations

stepancheg committed:
core.operators: remove Sequence? restriction on push

Because OutputStream is not a Sequence, and OutputStream needs push.
Probably we need something like Pushable.

stepancheg commented:
About `write`.
- I believe that there are filesystems and conditions when `write` does partial write.
- Sockets may also implement InputStream and OutputStream protocols.
- This change is harmless to users, because `writeBytes` returns number of bytes written, but `write` function don't. Users still don't have to check write result. Users don't need to call `writeBytes` explicitly unless they know what are they doing. Added [clarifying comment](https://github.com/stepancheg/clay/commit/9d60ce113ce00125429b6f50a0212e7ebf6a6150#L4R8).

`InputStream` and `OutputStream` now implement `Iterator` and `push` protocols correspondingly.

One important addition. I think that BufferedOutputStream must not flush buffer on destroy. Motivation is in [code comments](https://github.com/stepancheg/clay/commit/4df687e9994232beb9c7cea81a9cf72f66606bef#L0R49). So BufferOutputStream as well as File don't. Users need to explicitly call flush when using these API.

stdout and stderr are [flushed on exit](https://github.com/stepancheg/clay/commit/45964968481ac13373388288a53446d8101d4a7a#L1R7).
